design/StableRegions.md

```diff
@@ -0,0 +1,358 @@
+     in that table, if need be (see `rebuild`).
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region table
+
+ - purpose:
+   - relate a region ID to its size in pages.
+
+   - NB: This table exists as a "stable backup" of the per-region fields
+     that exist in the set of region objects, except for the access vectors' elements (see `block-region` table for their data).
+
+ - 32768 entries (statically sized).
+ - 8 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; }`
+ - the location of each entry in the table gives its corresponding region ID.
+ - the `size_in_pages` field measures the size of the region in terms of _pages_, not blocks.
+ - to do: track bit per region of whether in-use or reclaimed-and-available, somehow.
+
+
+### Overview of `rebuild`
+
+When upgrades work as expected, stable `Regions` are serialized and deserialized just like other stable data.
+
+For disaster recovery, we can **also** rebuild all of the region objects from data in stable memory.
+
+We use the `block-region` and `region` tables in stable memory to rebuild the regions' objects:
+
+ - The `region` table gives a size for each region, saying how large each vector for each region needs to be.
+ - The `block-region` table gives a relative position and region ID for each block ID.
+
+Once each regions' vectors have been allocated, the block-region table says how to fill them, one entry at a time.
+Unlike the Rust design, vector entries can be populated out-of-order.
+
+
+### Special (reserved) regions
+
+  - Region 0 -- Anonymous region, for supporting the legacy API that we have today, which lacks `Region` values.
+  - Region 1 -- "Reclaimed blocks" region that consists of blocks reclaimed from GC'd regions.
+  - Regions 2-15 -- Future use by Motoko RTS (TBD).
+
+### Overview of GC support (future work)
+
+- Regions are reprensented (see special subsection) with heap objects that are `stable`, but mutable.
+- They have special GC headers to recognize their special structure.
+- The Region table keeps track of which available IDs are in use as Region heap values are GC'd.
+
+#### Region allocation, re-allocation
+
+A region with a particular ID is first initially allocated by a global, bumped counter reaching it.
+
+When that happens, the region table is marked (TO DO) indicating that region is in use.
+
+The region heap object with that ID witnesses its lifetime, and is reclaimed by GC when it becomes unreachable.
+
+When the bump counter reaches its maximum value, the allocation algorithm changes to scanning the regions,
+looking for the first available one (with minimum ID).
+
+#### Region reclamation
+
+A heap object with a certain region ID lives until GC reclaims it, where the region table is re-marked (TO DO) to indicate that the region is free again.
```

Comment:
Just a detail (related to the previous comment): It could also be implicitly reclaimed if the stable region object does not survive upgrade, because it was garbage at the time of upgrade but was not yet detected as such by the GC.
Therefore, the GC would rather re-identify all live regions.